File: crates/rash-spec/src/types/ast.rs
use indexmap::IndexMap;
use serde::{Deserialize, Serialize};

use super::common::{Language, Tier, TypeRef};

/// AST Node — all handler logic is represented as a tree of these nodes.
/// Uses internal tagging: `#[serde(tag = "type")]`
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum AstNode {
    // ── Statements (Tier 0) ──

    /// Variable declaration: `let name = value`
    LetStatement {
        tier: Tier,
        name: String,
        #[serde(rename = "valueType", skip_serializing_if = "Option::is_none")]
        value_type: Option<TypeRef>,
        value: Box<AstNode>,
    },

    /// Variable assignment: `target = value`
    AssignStatement {
        tier: Tier,
        target: Box<AstNode>,
        value: Box<AstNode>,
    },

    /// Return statement
    ReturnStatement {
        tier: Tier,
        value: Box<AstNode>,
    },

    /// If/else conditional
    IfStatement {
        tier: Tier,
        condition: Box<AstNode>,
        then: Vec<AstNode>,
        #[serde(rename = "else", skip_serializing_if = "Option::is_none")]
        else_branch: Option<Vec<AstNode>>,
    },

    /// For-in loop
    ForStatement {
        tier: Tier,
        binding: String,
        iterable: Box<AstNode>,
        body: Vec<AstNode>,
    },

    /// While loop
    WhileStatement {
        tier: Tier,
        condition: Box<AstNode>,
        body: Vec<AstNode>,
    },

    /// Pattern matching
    MatchStatement {
        tier: Tier,
        expr: Box<AstNode>,
        arms: Vec<MatchArm>,
    },

    /// Try/catch/finally
    TryCatchStatement {
        tier: Tier,
        #[serde(rename = "try")]
        try_block: Vec<AstNode>,
        #[serde(rename = "catch")]
        catch_block: CatchClause,
        #[serde(rename = "finally", skip_serializing_if = "Option::is_none")]
        finally_block: Option<Vec<AstNode>>,
    },

    /// Throw an error
    ThrowStatement {
        tier: Tier,
        value: Box<AstNode>,
    },

    /// Expression used as statement
    ExpressionStatement {
        tier: Tier,
        expr: Box<AstNode>,
    },

    // ── Expressions (Tier 0) ──

    /// Literal value (string, number, boolean, null)
    Literal {
        tier: Tier,
        value: serde_json::Value,
    },

    /// Variable reference
    Identifier {
        tier: Tier,
        name: String,
    },

    /// Binary operation: `left op right`
    BinaryExpr {
        tier: Tier,
        operator: String,
        left: Box<AstNode>,
        right: Box<AstNode>,
    },

    /// Unary operation: `op operand`
    UnaryExpr {
        tier: Tier,
        operator: String,
        operand: Box<AstNode>,
    },

    /// Function call: `callee(args...)`
    CallExpr {
        tier: Tier,
        callee: Box<AstNode>,
        args: Vec<AstNode>,
    },

    /// Member access: `object.property`
    MemberExpr {
        tier: Tier,
        object: Box<AstNode>,
        property: String,
    },

    /// Index access: `object[index]`
    IndexExpr {
        tier: Tier,
        object: Box<AstNode>,
        index: Box<AstNode>,
    },

    /// Object literal: `{ key: value, ... }`
    ObjectExpr {
        tier: Tier,
        properties: IndexMap<String, AstNode>,
    },

    /// Array literal: `[elem, ...]`
    ArrayExpr {
        tier: Tier,
        elements: Vec<AstNode>,
    },

    /// Arrow/anonymous function
    ArrowFn {
        tier: Tier,
        params: Vec<String>,
        body: Vec<AstNode>,
    },

    /// Await expression
    AwaitExpr {
        tier: Tier,
        expr: Box<AstNode>,
    },

    /// Pipe expression: `a |> b |> c`
    PipeExpr {
        tier: Tier,
        stages: Vec<AstNode>,
    },

    /// Template string: `Hello ${name}`
    TemplateString {
        tier: Tier,
        parts: Vec<TemplatePart>,
    },

    // ── Domain Nodes (Tier 1) ──

    /// Database query (findOne, findMany, count, etc.)
    DbQuery {
        tier: Tier,
        model: String,
        operation: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        r#where: Option<serde_json::Value>,
        #[serde(rename = "orderBy", skip_serializing_if = "Option::is_none")]
        order_by: Option<serde_json::Value>,
        #[serde(skip_serializing_if = "Option::is_none")]
        skip: Option<Box<AstNode>>,
        #[serde(skip_serializing_if = "Option::is_none")]
        take: Option<Box<AstNode>>,
        #[serde(skip_serializing_if = "Option::is_none")]
        select: Option<Vec<String>>,
        #[serde(skip_serializing_if = "Option::is_none")]
        include: Option<serde_json::Value>,
    },

    /// Database mutation (insert, update, delete)
    DbMutate {
        tier: Tier,
        model: String,
        operation: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        data: Option<Box<AstNode>>,
        #[serde(skip_serializing_if = "Option::is_none")]
        r#where: Option<serde_json::Value>,
    },

    /// HTTP response
    HttpRespond {
        tier: Tier,
        status: u16,
        #[serde(skip_serializing_if = "Option::is_none")]
        headers: Option<IndexMap<String, AstNode>>,
        #[serde(skip_serializing_if = "Option::is_none")]
        body: Option<Box<AstNode>>,
    },

    /// Request context access (params, query, body, headers)
    CtxGet {
        tier: Tier,
        path: String,
    },

    /// Schema validation
    Validate {
        tier: Tier,
        schema: String,
        data: Box<AstNode>,
    },

    /// Password hashing
    HashPassword {
        tier: Tier,
        input: Box<AstNode>,
        algorithm: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        rounds: Option<u32>,
    },

    /// Password verification
    VerifyPassword {
        tier: Tier,
        password: Box<AstNode>,
        hash: Box<AstNode>,
        algorithm: String,
    },

    /// JWT signing
    SignToken {
        tier: Tier,
        payload: Box<AstNode>,
        #[serde(skip_serializing_if = "Option::is_none")]
        secret: Option<Box<AstNode>>,
        #[serde(skip_serializing_if = "Option::is_none")]
        options: Option<serde_json::Value>,
    },

    /// JWT verification
    VerifyToken {
        tier: Tier,
        token: Box<AstNode>,
        #[serde(skip_serializing_if = "Option::is_none")]
        secret: Option<Box<AstNode>>,
    },

    /// Send email
    SendEmail {
        tier: Tier,
        to: Box<AstNode>,
        subject: Box<AstNode>,
        body: Box<AstNode>,
    },

    /// Emit event
    EmitEvent {
        tier: Tier,
        event: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        data: Option<Box<AstNode>>,
    },

    /// Log message
    LogMessage {
        tier: Tier,
        level: String,
        message: Box<AstNode>,
    },

    // ── Bridge Node (Tier 3) ──

    /// Native language bridge — locks to a specific language
    NativeBridge {
        tier: Tier,
        language: Language,
        package: String,
        import: NativeBridgeImport,
        call: NativeBridgeCall,
        #[serde(rename = "returnType", skip_serializing_if = "Option::is_none")]
        return_type: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        fallback: Option<NativeBridgeFallback>,
    },
}

/// Match arm for MatchStatement
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MatchArm {
    pub pattern: AstNode,
    pub body: Vec<AstNode>,
}

/// Catch clause for TryCatchStatement
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CatchClause {
    pub binding: String,
    pub body: Vec<AstNode>,
}

/// Template string part
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum TemplatePart {
    #[serde(rename = "text")]
    Text { value: String },
    #[serde(rename = "expr")]
    Expr { value: Box<AstNode> },
}

/// NativeBridge import declaration
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct NativeBridgeImport {
    pub name: String,
    pub from: String,
}

/// NativeBridge function call
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct NativeBridgeCall {
    pub method: String,
    pub args: Vec<AstNode>,
}

/// NativeBridge fallback node
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct NativeBridgeFallback {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub node: Box<AstNode>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_let_statement_roundtrip() {
        let json = serde_json::json!({
            "type": "LetStatement",
            "tier": 0,
            "name": "userId",
            "value": {
                "type": "CtxGet",
                "tier": 1,
                "path": "params.id"
            }
        });

        let node: AstNode = serde_json::from_value(json.clone()).unwrap();
        match &node {
            AstNode::LetStatement { name, tier, .. } => {
                assert_eq!(name, "userId");
                assert_eq!(*tier, Tier::Universal);
            }
            _ => panic!("Expected LetStatement"),
        }

        // Roundtrip
        let serialized = serde_json::to_value(&node).unwrap();
        let node2: AstNode = serde_json::from_value(serialized).unwrap();
        assert_eq!(node, node2);
    }

    #[test]
    fn test_if_statement_with_nested_nodes() {
        let json = serde_json::json!({
            "type": "IfStatement",
            "tier": 0,
            "condition": {
                "type": "BinaryExpr",
                "tier": 0,
                "operator": "==",
                "left": { "type": "Identifier", "tier": 0, "name": "user" },
                "right": { "type": "Literal", "tier": 0, "value": null }
            },
            "then": [
                {
                    "type": "ReturnStatement",
                    "tier": 0,
                    "value": {
                        "type": "HttpRespond",
                        "tier": 1,
                        "status": 404,
                        "body": {
                            "type": "ObjectExpr",
                            "tier": 0,
                            "properties": {
                                "message": { "type": "Literal", "tier": 0, "value": "User not found" },
                                "code": { "type": "Literal", "tier": 0, "value": "NOT_FOUND" }
                            }
                        }
                    }
                }
            ],
            "else": null
        });

        let node: AstNode = serde_json::from_value(json).unwrap();
        match &node {
            AstNode::IfStatement { then, .. } => {
                assert_eq!(then.len(), 1);
            }
            _ => panic!("Expected IfStatement"),
        }
    }

    #[test]
    fn test_db_query_deserialization() {
        let json = serde_json::json!({
            "type": "DbQuery",
            "tier": 1,
            "model": "User",
            "operation": "findUnique",
            "where": {
                "id": {
                    "type": "Identifier",
                    "tier": 0,
                    "name": "userId"
                }
            }
        });

        let node: AstNode = serde_json::from_value(json).unwrap();
        match &node {
            AstNode::DbQuery {
                model, operation, ..
            } => {
                assert_eq!(model, "User");
                assert_eq!(operation, "findUnique");
            }
            _ => panic!("Expected DbQuery"),
        }
    }

    #[test]
    fn test_native_bridge_deserialization() {
        let json = serde_json::json!({
            "type": "NativeBridge",
            "tier": 3,
            "language": "typescript",
            "package": "npm:bcrypt",
            "import": { "name": "bcrypt", "from": "bcrypt" },
            "call": {
                "method": "bcrypt.hash",
                "args": [
                    { "type": "Identifier", "tier": 0, "name": "password" },
                    { "type": "Literal", "tier": 0, "value": 10 }
                ]
            },
            "returnType": "string",
            "fallback": {
                "description": "다른 언어로 변환 시 대체 노드",
                "node": {
                    "type": "HashPassword",
                    "tier": 1,
                    "input": { "type": "Identifier", "tier": 0, "name": "password" },
                    "algorithm": "bcrypt",
                    "rounds": 10
                }
            }
        });

        let node: AstNode = serde_json::from_value(json).unwrap();
        match &node {
            AstNode::NativeBridge {
                tier,
                language,
                package,
                fallback,
                ..
            } => {
                assert_eq!(*tier, Tier::Bridge);
                assert_eq!(*language, Language::Typescript);
                assert_eq!(package, "npm:bcrypt");
                assert!(fallback.is_some());
            }
            _ => panic!("Expected NativeBridge"),
        }
    }

    #[test]
    fn test_full_handler_body_from_docs() {
        let json = serde_json::json!([
            {
                "type": "LetStatement",
                "tier": 0,
                "name": "userId",
                "value": {
                    "type": "CtxGet",
                    "tier": 1,
                    "path": "params.id"
                }
            },
            {
                "type": "LetStatement",
                "tier": 0,
                "name": "user",
                "value": {
                    "type": "AwaitExpr",
                    "tier": 1,
                    "expr": {
                        "type": "DbQuery",
                        "tier": 1,
                        "model": "User",
                        "operation": "findUnique",
                        "where": {
                            "id": { "type": "Identifier", "tier": 0, "name": "userId" }
                        }
                    }
                }
            },
            {
                "type": "IfStatement",
                "tier": 0,
                "condition": {
                    "type": "BinaryExpr",
                    "tier": 0,
                    "operator": "==",
                    "left": { "type": "Identifier", "tier": 0, "name": "user" },
                    "right": { "type": "Literal", "tier": 0, "value": null }
                },
                "then": [
                    {
                        "type": "ReturnStatement",
                        "tier": 0,
                        "value": {
                            "type": "HttpRespond",
                            "tier": 1,
                            "status": 404,
                            "body": {
                                "type": "ObjectExpr",
                                "tier": 0,
                                "properties": {
                                    "message": { "type": "Literal", "tier": 0, "value": "User not found" },
                                    "code": { "type": "Literal", "tier": 0, "value": "NOT_FOUND" }
                                }
                            }
                        }
                    }
                ],
                "else": null
            },
            {
                "type": "ReturnStatement",
                "tier": 0,
                "value": {
                    "type": "HttpRespond",
                    "tier": 1,
                    "status": 200,
                    "body": { "type": "Identifier", "tier": 0, "name": "user" }
                }
            }
        ]);

        let body: Vec<AstNode> = serde_json::from_value(json).unwrap();
        assert_eq!(body.len(), 4);

        // Verify roundtrip
        let serialized = serde_json::to_value(&body).unwrap();
        let body2: Vec<AstNode> = serde_json::from_value(serialized).unwrap();
        assert_eq!(body, body2);
    }
}

File: src/utils/ast.ts
import type { AstNode, Tier } from "../ipc/types";

let nextId = 1;
function uid(): string {
  return `node_${Date.now()}_${nextId++}`;
}

// ---------------------------------------------------------------------------
// Generic factory
// ---------------------------------------------------------------------------

export function createNode(type: string, tier: Tier, extra: Record<string, unknown> = {}): AstNode {
  return { id: uid(), type, tier, ...extra };
}

// ---------------------------------------------------------------------------
// Statement factories
// ---------------------------------------------------------------------------

export function createLetStatement(name = "", value?: AstNode): AstNode {
  return createNode("LetStatement", 0, { name, value: value ?? null, mutable: false });
}

export function createIfStatement(condition?: AstNode, body: AstNode[] = [], elseBody: AstNode[] = []): AstNode {
  return createNode("IfStatement", 0, { condition: condition ?? null, body, elseBody });
}

export function createForStatement(variable = "item", iterable?: AstNode, body: AstNode[] = []): AstNode {
  return createNode("ForStatement", 0, { variable, iterable: iterable ?? null, body });
}

export function createWhileStatement(condition?: AstNode, body: AstNode[] = []): AstNode {
  return createNode("WhileStatement", 0, { condition: condition ?? null, body });
}

export function createReturnStatement(value?: AstNode): AstNode {
  return createNode("ReturnStatement", 0, { value: value ?? null });
}

export function createTryCatchStatement(
  tryBody: AstNode[] = [],
  catchParam = "err",
  catchBody: AstNode[] = [],
): AstNode {
  return createNode("TryCatchStatement", 0, { tryBody, catchParam, catchBody });
}

export function createThrowStatement(value?: AstNode): AstNode {
  return createNode("ThrowStatement", 0, { value: value ?? null });
}

export function createMatchStatement(subject?: AstNode, arms: { pattern: string; body: AstNode[] }[] = []): AstNode {
  return createNode("MatchStatement", 0, { subject: subject ?? null, arms });
}

// ---------------------------------------------------------------------------
// Expression factories
// ---------------------------------------------------------------------------

export function createLiteral(value: unknown = ""): AstNode {
  return createNode("Literal", 0, { value });
}

export function createIdentifier(name = ""): AstNode {
  return createNode("Identifier", 0, { name });
}

export function createBinaryExpression(operator = "+", left?: AstNode, right?: AstNode): AstNode {
  return createNode("BinaryExpression", 0, { operator, left: left ?? null, right: right ?? null });
}

export function createCallExpression(callee?: AstNode, args: AstNode[] = []): AstNode {
  return createNode("CallExpression", 0, { callee: callee ?? null, arguments: args });
}

export function createMemberExpression(object?: AstNode, property = ""): AstNode {
  return createNode("MemberExpression", 0, { object: object ?? null, property });
}

export function createObjectExpression(properties: { key: string; value: AstNode }[] = []): AstNode {
  return createNode("ObjectExpression", 0, { properties });
}

export function createArrayExpression(elements: AstNode[] = []): AstNode {
  return createNode("ArrayExpression", 0, { elements });
}

export function createTemplateLiteral(parts: (string | AstNode)[] = []): AstNode {
  return createNode("TemplateLiteral", 0, { parts });
}

// ---------------------------------------------------------------------------
// Domain (Tier 1) factories
// ---------------------------------------------------------------------------

export function createDbQuery(model = "", operation = "findMany", where?: AstNode): AstNode {
  return createNode("DbQuery", 1, { model, operation, where: where ?? null });
}

export function createDbMutate(model = "", operation = "create", data?: AstNode): AstNode {
  return createNode("DbMutate", 1, { model, operation, data: data ?? null });
}

export function createHttpRespond(status = 200, body?: AstNode, headers?: AstNode): AstNode {
  return createNode("HttpRespond", 1, { status, body: body ?? null, headers: headers ?? null });
}

export function createCtxGet(key = ""): AstNode {
  return createNode("CtxGet", 1, { key });
}

export function createValidate(schema = "", value?: AstNode): AstNode {
  return createNode("Validate", 1, { schema, value: value ?? null });
}

export function createHashPassword(value?: AstNode): AstNode {
  return createNode("HashPassword", 1, { value: value ?? null });
}

export function createSignToken(payload?: AstNode, secret?: AstNode): AstNode {
  return createNode("SignToken", 1, { payload: payload ?? null, secret: secret ?? null });
}

// ---------------------------------------------------------------------------
// Bridge (Tier 3)
// ---------------------------------------------------------------------------

export function createNativeBridge(language = "typescript", module = "", method = "", args: AstNode[] = []): AstNode {
  return createNode("NativeBridge", 3, { language, module, method, arguments: args });
}

// ---------------------------------------------------------------------------
// Tree utilities
// ---------------------------------------------------------------------------

export function getNodeTier(node: AstNode): Tier {
  return node.tier;
}

/**
 * Remove a node by id from an array of nodes (searches recursively).
 * Returns a new array (immutable).
 */
export function removeNode(nodes: AstNode[], nodeId: string): AstNode[] {
  return nodes
    .filter((n) => (n as Record<string, unknown>).id !== nodeId)
    .map((n) => removeNodeFromChildren(n, nodeId));
}

function removeNodeFromChildren(node: AstNode, nodeId: string): AstNode {
  const result: Record<string, unknown> = { ...node };
  for (const key of Object.keys(result)) {
    const val = result[key];
    if (Array.isArray(val) && val.length > 0 && val[0] && typeof val[0] === "object" && "type" in val[0]) {
      result[key] = removeNode(val as AstNode[], nodeId);
    } else if (val && typeof val === "object" && "type" in (val as Record<string, unknown>)) {
      if ((val as Record<string, unknown>).id === nodeId) {
        result[key] = null;
      }
    }
  }
  return result as AstNode;
}

/**
 * Insert a node into a parent's child array at position `index`.
 * `parentId` is the id of the parent node; the child array is searched by
 * looking at known container keys (body, tryBody, catchBody, elements, etc.).
 */
export function insertNode(
  nodes: AstNode[],
  parentId: string | null,
  node: AstNode,
  index?: number,
): AstNode[] {
  // Top-level insert
  if (parentId === null) {
    const idx = index ?? nodes.length;
    const copy = [...nodes];
    copy.splice(idx, 0, node);
    return copy;
  }

  return nodes.map((n) => {
    if ((n as Record<string, unknown>).id === parentId) {
      return insertIntoFirstChildArray(n, node, index);
    }
    return insertNodeRecursive(n, parentId, node, index);
  });
}

const CONTAINER_KEYS = ["body", "tryBody", "catchBody", "elseBody", "elements", "arguments", "properties", "arms"];

function insertIntoFirstChildArray(parent: AstNode, node: AstNode, index?: number): AstNode {
  const result: Record<string, unknown> = { ...parent };
  for (const key of CONTAINER_KEYS) {
    if (Array.isArray(result[key])) {
      const arr = [...(result[key] as AstNode[])];
      const idx = index ?? arr.length;
      arr.splice(idx, 0, node);
      result[key] = arr;
      return result as AstNode;
    }
  }
  return result as AstNode;
}

function insertNodeRecursive(node: AstNode, parentId: string, child: AstNode, index?: number): AstNode {
  const result: Record<string, unknown> = { ...node };
  for (const key of Object.keys(result)) {
    const val = result[key];
    if (Array.isArray(val) && val.length > 0 && val[0] && typeof val[0] === "object" && "type" in val[0]) {
      result[key] = insertNode(val as AstNode[], parentId, child, index);
    }
  }
  return result as AstNode;
}

/**
 * Update a single property on a node found by its `id`.
 */
export function updateNodeProperty(nodes: AstNode[], nodeId: string, key: string, value: unknown): AstNode[] {
  return nodes.map((n) => {
    if ((n as Record<string, unknown>).id === nodeId) {
      return { ...n, [key]: value };
    }
    const result: Record<string, unknown> = { ...n };
    for (const k of Object.keys(result)) {
      const v = result[k];
      if (Array.isArray(v) && v.length > 0 && v[0] && typeof v[0] === "object" && "type" in v[0]) {
        result[k] = updateNodeProperty(v as AstNode[], nodeId, key, value);
      }
    }
    return result as AstNode;
  });
}
